File: mobile-app/lib/utils/security.dart
/// Utilitaires de sécurité pour l'application mobile
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// Valider un email
bool isValidEmail(String email) {
  final emailRegex = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$');
  return emailRegex.hasMatch(email);
}

/// Valider un mot de passe fort
bool isValidPassword(String password) {
  // Au moins 8 caractères, une majuscule, une minuscule, un chiffre
  final passwordRegex = RegExp(r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$');
  return passwordRegex.hasMatch(password);
}

/// Valider et nettoyer un nom de fichier
String sanitizeFileName(String fileName) {
  // Supprimer les caractères dangereux
  return fileName
      .replaceAll(RegExp(r'[<>:"/\\|?*\x00-\x1f]'), '')
      .replaceAll('..', '')
      .trim()
      .substring(0, fileName.length > 255 ? 255 : fileName.length);
}

/// Échapper les caractères HTML
String escapeHtml(String text) {
  return text
      .replaceAll('&', '&amp;')
      .replaceAll('<', '&lt;')
      .replaceAll('>', '&gt;')
      .replaceAll('"', '&quot;')
      .replaceAll("'", '&#039;');
}

/// Générer un hash SHA-256
String generateHash(String input) {
  final bytes = utf8.encode(input);
  final digest = sha256.convert(bytes);
  return digest.toString();
}

// SecureStorage a été déplacé dans services/secure_storage.dart
// Importez-le avec: import '../services/secure_storage.dart';

/// Valider un chemin de fichier pour prévenir les directory traversal
bool validateFilePath(String filePath, String baseDir) {
  final normalizedPath = filePath.replaceAll('..', '');
  return normalizedPath.startsWith(baseDir);
}

File: frontend-web/src/utils/security.js
/**
 * Utilitaires de sécurité pour le frontend
 */

/**
 * Sanitize une chaîne de caractères pour prévenir XSS
 */
export function sanitizeString(str) {
  if (typeof str !== 'string') return '';
  
  const div = document.createElement('div');
  div.textContent = str;
  return div.innerHTML;
}

/**
 * Valider un email
 */
export function isValidEmail(email) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Valider un mot de passe fort
 */
export function isValidPassword(password) {
  // Au moins 8 caractères, une majuscule, une minuscule, un chiffre
  const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$/;
  return passwordRegex.test(password);
}

/**
 * Échapper les caractères HTML
 */
export function escapeHtml(text) {
  const map = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#039;'
  };
  return text.replace(/[&<>"']/g, m => map[m]);
}

/**
 * Valider et nettoyer un nom de fichier
 */
export function sanitizeFileName(fileName) {
  // Supprimer les caractères dangereux
  return fileName
    .replace(/[<>:"/\\|?*]/g, '')
    .replace(/\.\./g, '')
    .trim()
    .substring(0, 255); // Limiter la longueur
}

/**
 * Générer un token CSRF
 */
export function generateCSRFToken() {
  const array = new Uint8Array(32);
  crypto.getRandomValues(array);
  return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
}

/**
 * Stocker un token de manière sécurisée
 */
export function setSecureToken(key, value) {
  try {
    localStorage.setItem(key, value);
  } catch (e) {
    console.error('Failed to store token:', e);
  }
}

/**
 * Récupérer un token de manière sécurisée
 */
export function getSecureToken(key) {
  try {
    return localStorage.getItem(key);
  } catch (e) {
    console.error('Failed to retrieve token:', e);
    return null;
  }
}

/**
 * Supprimer un token de manière sécurisée
 */
export function removeSecureToken(key) {
  try {
    localStorage.removeItem(key);
  } catch (e) {
    console.error('Failed to remove token:', e);
  }
}





